day5/main.go

```diff
@@ -0,0 +1,269 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+	"unicode"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Println("input file path not provided")
+		return
+	}
+	input := readFileLineByLine(os.Args[1])
+	fmt.Println("answer for part1: ", createMap(input))
+	fmt.Println("answer for part2: ", part2(input))
+}
+
+func readFileLineByLine(path string) []string {
+	file, err := os.Open(path)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer file.Close()
+
+	var output []string
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		output = append(output, scanner.Text())
+	}
+
+	if err := scanner.Err(); err != nil {
+		log.Fatal(err)
+	}
+
+	return output
+}
+
+type zoo struct {
+	end        int
+	startValue int
+}
+
+func part2(input []string) int {
+	seeds := make(map[int]int)
+	seedSoil := make(map[int]zoo)
+	soilFert := make(map[int]zoo)
+	fertWater := make(map[int]zoo)
+	waterLight := make(map[int]zoo)
+	lightTemp := make(map[int]zoo)
+	tempHum := make(map[int]zoo)
+	humLoc := make(map[int]zoo)
+
+	var mapPointer map[int]zoo
+	for _, line := range input {
+		if strings.Index(line, ":") != -1 && line[:strings.Index(line, ":")] == "seeds" {
+			inputSeeds := fetchNumsInLine(line[strings.Index(line, ":"):])
+			for i := 0; i < len(inputSeeds); i++ {
+				seeds[inputSeeds[i]] = inputSeeds[i+1] + inputSeeds[i] - 1
+				i++
+			}
+			continue
+		}
+
+		if strings.Index(line, "seed-to-soil") != -1 {
+			mapPointer = seedSoil
+			continue
+		}
+
+		if strings.Index(line, "soil-to-fertilizer") != -1 {
+			mapPointer = soilFert
+			continue
+		}
+
+		if strings.Index(line, "fertilizer-to-water") != -1 {
+			mapPointer = fertWater
+			continue
+		}
+
+		if strings.Index(line, "water-to-light") != -1 {
+			mapPointer = waterLight
+			continue
+		}
+
+		if strings.Index(line, "light-to-temperature") != -1 {
+			mapPointer = lightTemp
+			continue
+		}
+
+		if strings.Index(line, "temperature-to-humidity") != -1 {
+			mapPointer = tempHum
+			continue
+		}
+
+		if strings.Index(line, "humidity-to-location") != -1 {
+			mapPointer = humLoc
+			continue
+		}
+
+		nums := fetchNumsInLine(line)
+		if len(nums) == 3 {
+			valueStart := nums[0]
+			keyStart := nums[1]
+			count := nums[2] - 1
+
+			mapPointer[valueStart] = zoo{
+				end:        valueStart + count,
+				startValue: keyStart,
+			}
+		}
+
+	}
+	i := 0
+	for ; ; i++ {
+
+		hum := findValue(humLoc, &i)
+		temp := findValue(tempHum, hum)
+		light := findValue(lightTemp, temp)
+		water := findValue(waterLight, light)
+		fert := findValue(fertWater, water)
+		soil := findValue(soilFert, fert)
+		seed := findValue(seedSoil, soil)
+
+		for k, v := range seeds {
```

Comment:
I think using a binary search (while having sorted the seeds first) could be an improvement.

Reply:
Could maybe work if the result of findValue received a list of ranges.. and returned a new set of ranges. I'm sure there will be little overlap with each stage. But it could be a good improvement.

Reply:
> I think using a binary search (while having sorted the seeds first) could be an improvement.

@teivah Yeah, that would definitely be an improvement. `seeds` type is `map`, not a `slice` and so sorting would not work as is. But again, having a map of seeds is also not adding much value. I'll change the type and use binary search 👍 

> Could maybe work if the result of findValue received a list of ranges.. and returned a new set of ranges. I'm sure there will be little overlap with each stage. But it could be a good improvement.

💯 

Will try out both suggested improvements later today, thanks folks :)